src/RulesetEngine.php

```diff
@@ -83,7 +102,12 @@ public function process(
             }
         }
 
-        return new Context($rules, $errors, $warnings);
+        $context = new Context($rules, $errors, $warnings);
+        $event = new PostRulesetProcessingEvent($context);
+
+        $this->eventDispatcher->dispatch($event);
+
+        return $event->getContext();
```

Comment:
Idea 2: Next best thing to dispatching events for each ruleset action

We only emit a single event after all rules are collected and we can then replace the whole context

Cons:
 - Altering a single violation requires tedious searching through the whole context
 - Replacing the whole context opens up possibility for errors and confusing output

Pros:
 - Opens up more possibilities to users acting on the event
 - Single event means minimal overhead for Deptrac

Reply:
I like this one much better, to be honest. But to improve it, I would like to see the event also having part of the configuration like layer/ruleset definition. So that users can define their own `RulesetEngine` to some extent.